checker/match.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2018 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"fmt"
+	"regexp"
+
+	authz "github.com/envoyproxy/data-plane-api/api/auth"
+
+	"github.com/projectcalico/app-policy/proto"
+	"github.com/projectcalico/libcalico-go/lib/selector"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const SPIFFE_ID_PATTERN = "^spiffe://[^/]+/ns/([^/]+)/sa/([^/]+)$"
+
+var spiffeIdRegExp *regexp.Regexp
+
+// match checks if the Rule matches the request.  It returns true if the Rule matches, false otherwise.
+func match(rule *proto.Rule, req *authz.CheckRequest) bool {
+	log.Debugf("Checking rule %v on request %v", rule, req)
+	attr := req.GetAttributes()
+	return matchSource(rule, attr.GetSource()) && matchRequest(rule, attr.GetRequest())
+}
+
+func matchSource(r *proto.Rule, peer *authz.AttributeContext_Peer) bool {
+	// TODO IPSets
+	return matchServiceAccounts(r.GetSrcServiceAccount(), peer)
+}
+
+func matchRequest(rule *proto.Rule, req *authz.AttributeContext_Request) bool {
+	log.WithFields(log.Fields{
+		"request": req,
+	}).Debug("Matching request.")
+	return matchHTTP(rule.GetHttp(), req.GetHttp())
+}
+
+func matchServiceAccounts(saMatch *proto.ServiceAccountSelector, peer *authz.AttributeContext_Peer) bool {
+	principle := peer.GetPrincipal()
+	labels := peer.GetLabels()
+	log.WithFields(log.Fields{
+		"peer":   principle,
+		"labels": labels,
+		"rule":   saMatch},
+	).Debug("Matching service account.")
+	accountName, namespace, err := parseSpiffeId(principle)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"principle": principle,
+			"msg":       err,
+		}).Warn("Unable to parse authenticated principle as SPIFFE ID.")
+		return false
```

Comment:
Returning `false` so no match. Could this potentially result in bypassing a `serviceAccount` match and therefore skipping a rule and moving on to the next one? Could we then hit a rule which has `Allow`?
Should we take/return the default action in this case? (By default I mean the default-deny/allow, when no rule matches).

Reply:
Yes, it can result in bypassing a serviceAccount match and hitting some later matching `Allow` rule.

I'm sure sure what the right behavior should be if the SPIFFE ID is malformed. One alternative would be to check that it is parsable before even attempting to apply policy, and reject any requests with a malformed ID. What do you think? I'll also ask Alex

Reply:
Actually, I take this back. This code gets executed regardless of whether there is a service account selector on the rule or not, so a malformed SPIFFE ID means that no rules will match and we'll have a default deny.